kernel/rros/observable.rs

```diff
@@ -0,0 +1,990 @@
+        // `data` is initialized with a length,
+        // this length is the length which the caller want to read.
+        rros_read_observable(
+            unsafe { observable.as_mut().unwrap() },
+            data,
+            data.len(),
+            file.is_blocking(),
+        )
+    }
+
+    fn oob_read<T: IoBufferWriter>(_this: &CloneData, file: &File, data: &mut T) -> Result<usize> {
+        pr_info!("I'm the oob_read ops of the observable factory.");
+        let fbind: *const RrosFileBinding =
+            unsafe { (*(file.get_ptr())).private_data as *const RrosFileBinding };
+        let observable = unsafe { (*(*fbind).element).pointer as *mut RrosObservable };
+        // `data` is initialized with a length,
+        // this length is the length which the caller want to read.
+        rros_read_observable(
+            unsafe { observable.as_mut().unwrap() },
+            data,
+            data.len(),
+            file.is_blocking(),
+        )
+    }
+
+    fn write<T: kernel::io_buffer::IoBufferReader>(
+        _this: &CloneData,
+        file: &File,
+        data: &mut T,
+        _offset: u64,
+    ) -> Result<usize> {
+        pr_info!("I'm the write ops of the observable factory.");
+        let fbind: *const RrosFileBinding =
+            unsafe { (*(file.get_ptr())).private_data as *const RrosFileBinding };
+        let observable = unsafe { (*(*fbind).element).pointer as *mut RrosObservable };
+
+        rros_write_observable(unsafe { observable.as_mut().unwrap() }, data, data.len())
+    }
+
+    fn oob_write<T: kernel::io_buffer::IoBufferReader>(
+        _this: &CloneData,
+        file: &File,
+        data: &mut T,
+    ) -> Result<usize> {
+        pr_info!("I'm the oob_write ops of the observable factory.");
+        let fbind: *const RrosFileBinding =
+            unsafe { (*(file.get_ptr())).private_data as *const RrosFileBinding };
+        let observable = unsafe { (*(*fbind).element).pointer as *mut RrosObservable };
+        // let observer_next = unsafe { &mut (*(*observable).observers.next).next as *mut _ as *mut RrosObserver };
+        // test_pending_list(unsafe { (*(observer_next)).pending_list });
+
+        rros_write_observable(unsafe { observable.as_mut().unwrap() }, data, data.len())
+    }
+
+    fn release(_this: Box<CloneData>, _file: &File) {
+        pr_info!("I'm the release ops of the observable factory.");
+    }
+}
+
+pub fn observable_factory_build(
+    fac: &'static mut SpinLock<RrosFactory>,
+    uname: &'static CStr,
+    _u_attrs: Option<*mut u8>,
+    clone_flags: i32,
+    _state_offp: &u32,
+) -> Rc<RefCell<RrosElement>> {
+    pr_info!("[observable] observable_factory_build: start");
+    // in case failed
+    if (clone_flags & !RROS_OBSERVABLE_CLONE_FLAGS) != 0 {
+        pr_info!("[observable] this is a wrong value");
+        // return Err(Error::EINVAL)
+    }
+
+    let observable = RrosObservable::new();
```

Comment:
The following part can have better implementation. Why not keeping `observable` in the Box and using rust to manage it?

Reply:
This is an issue that needs to be addressed: in EVL, when an observable element is allocated, the evl_index_factory_element function is invoked to insert the observable.element into a red-black tree. However, RROS currently lacks the implementation of the rros_index_factory_element function. If straightly using Box here, it would violate Rust’s ownership rules in some cases. As a temporary workaround, I can only achieve the corresponding functionality using raw pointers.